File: go/Assembler/engine/normaljson/engine_normal_json.go
package engine_normaljson

import (
	Logger "arshu/common"
	"assembler/common"
	interfaces "assembler/interface"
	"fmt"
	"strings"
)

type EngineNormalJson struct {
	AppViewPrefix string
}

func NewEngineNormalJson(prefix string) *EngineNormalJson {
	return &EngineNormalJson{AppViewPrefix: prefix}
}

func (e *EngineNormalJson) MergeTemplates(appSite, appFile, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	Logger.Debug(fmt.Sprintf("MergeTemplates called: appSite=%s, appFile=%s, appView=%s, enableJson=%t", appSite, appFile, appView, enableJsonProcessing), "EngineNormalJson")

	if loader == nil {
		Logger.Warn("No loader provided", "EngineNormalJson")
		return ""
	}

	result := loader.GetTemplateHtml(appSite, appFile, appView, e.AppViewPrefix)
	if result == nil {
		Logger.Warn(fmt.Sprintf("Main template not found for appSite=%s, appFile=%s", appSite, appFile), "EngineNormalJson")
		return ""
	}

	contentHtml, ok := result.(string)
	if !ok || contentHtml == "" {
		Logger.Warn(fmt.Sprintf("Main template not found or invalid type for appSite=%s, appFile=%s", appSite, appFile), "EngineNormalJson")
		return ""
	}

	Logger.Debug(fmt.Sprintf("Main template found, html size: %d", len(contentHtml)), "EngineNormalJson")

	if enableJsonProcessing {
		contentHtml = loader.MergeHtmlWithJson(contentHtml, appSite, appFile)
		Logger.Debug(fmt.Sprintf("After main JSON merge: %d chars", len(contentHtml)), "EngineNormalJson")
	}

	previous := ""
	maxPasses := 10
	actualPasses := 0
	for pass := 0; pass < maxPasses; pass++ {
		previous = contentHtml
		actualPasses = pass + 1

		Logger.Debug(fmt.Sprintf("Pass %d, current size: %d", actualPasses, len(contentHtml)), "EngineNormalJson")

		contentHtml = e.mergeTemplateSlots(contentHtml, appSite, appView, loader, enableJsonProcessing)
		Logger.Debug(fmt.Sprintf("After slot merge: %d chars", len(contentHtml)), "EngineNormalJson")

		contentHtml = e.replaceTemplatePlaceholders(contentHtml, appSite, appView, loader, enableJsonProcessing)
		Logger.Debug(fmt.Sprintf("After placeholder replacement: %d chars", len(contentHtml)), "EngineNormalJson")

		if contentHtml == previous {
			Logger.Debug(fmt.Sprintf("No changes in pass %d, stopping", actualPasses), "EngineNormalJson")
			break
		}
	}

	Logger.Debug(fmt.Sprintf("MergeTemplates complete after %d passes: output size=%d", actualPasses, len(contentHtml)), "EngineNormalJson")
	return contentHtml
}

func (e *EngineNormalJson) getTemplateWithJson(appSite, templateName, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	result := loader.GetTemplateHtml(appSite, templateName, appView, e.AppViewPrefix)
	if result == nil {
		return ""
	}

	html, ok := result.(string)
	if !ok || html == "" {
		return ""
	}

	Logger.Debug(fmt.Sprintf("GetTemplateWithJson: template=%s, html size=%d", templateName, len(html)), "EngineNormalJson")

	if enableJsonProcessing {
		originalSize := len(html)
		html = loader.MergeHtmlWithJson(html, appSite, templateName)
		Logger.Debug(fmt.Sprintf("After JSON merge for %s: size %d -> %d", templateName, originalSize, len(html)), "EngineNormalJson")
	}

	return html
}

func (e *EngineNormalJson) mergeTemplateSlots(contentHtml, appSite, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	if contentHtml == "" {
		return contentHtml
	}

	previous := ""
	for {
		previous = contentHtml
		contentHtml = e.processTemplateSlots(contentHtml, appSite, appView, loader, enableJsonProcessing)
		if contentHtml == previous {
			break
		}
	}
	return contentHtml
}

func (e *EngineNormalJson) processTemplateSlots(contentHtml, appSite, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	result := contentHtml
	searchPos := 0

	for searchPos < len(result) {
		openStart := strings.Index(result[searchPos:], "{{#")
		if openStart == -1 {
			break
		}
		openStart += searchPos

		openEnd := strings.Index(result[openStart+3:], "}}")
		if openEnd == -1 {
			break
		}
		openEnd += openStart + 3

		templateName := strings.TrimSpace(result[openStart+3 : openEnd])
		if templateName == "" || !common.IsAlphaNumeric(templateName) {
			searchPos = openStart + 1
			continue
		}

		closeTag := "{{/" + templateName + "}}"
		closeStart, found := common.FindMatchingCloseTag(result, openEnd+2, "{{#"+templateName+"}}", closeTag)
		if !found {
			searchPos = openStart + 1
			continue
		}

		innerStart := openEnd + 2
		innerContent := result[innerStart:closeStart]

		templateHtml := e.getTemplateWithJson(appSite, templateName, appView, loader, enableJsonProcessing)

		if templateHtml != "" {
			slotContents := e.extractSlotContents(innerContent, appSite, appView, loader, enableJsonProcessing)

			processedTemplate := templateHtml
			for key, value := range slotContents {
				processedTemplate = strings.ReplaceAll(processedTemplate, key, value)
			}

			processedTemplate = common.RemoveRemainingSlotPlaceholders(processedTemplate)

			fullMatch := result[openStart : closeStart+len(closeTag)]
			result = strings.ReplaceAll(result, fullMatch, processedTemplate)
			searchPos = openStart + len(processedTemplate)
		} else {
			searchPos = openStart + 1
		}
	}

	return result
}

func (e *EngineNormalJson) extractSlotContents(innerContent, appSite, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) map[string]string {
	slotContents := make(map[string]string)
	searchPos := 0

	for searchPos < len(innerContent) {
		slotStart := strings.Index(innerContent[searchPos:], "{{@HTMLPLACEHOLDER")
		if slotStart == -1 {
			break
		}
		slotStart += searchPos

		afterPlaceholder := slotStart + 18
		slotNum := ""
		pos := afterPlaceholder

		for pos < len(innerContent) && innerContent[pos] >= '0' && innerContent[pos] <= '9' {
			slotNum += string(innerContent[pos])
			pos++
		}

		if pos+1 >= len(innerContent) || innerContent[pos:pos+2] != "}}" {
			searchPos = slotStart + 1
			continue
		}

		slotOpenEnd := pos + 2

		closeTag := "{{/HTMLPLACEHOLDER}}"
		if slotNum != "" {
			closeTag = "{{/HTMLPLACEHOLDER" + slotNum + "}}"
		}
		openTag := "{{@HTMLPLACEHOLDER}}"
		if slotNum != "" {
			openTag = "{{@HTMLPLACEHOLDER" + slotNum + "}}"
		}

		closeStart, found := common.FindMatchingCloseTag(innerContent, slotOpenEnd, openTag, closeTag)
		if !found {
			searchPos = slotStart + 1
			continue
		}

		slotContent := innerContent[slotOpenEnd:closeStart]

		slotKey := "{{$HTMLPLACEHOLDER}}"
		if slotNum != "" {
			slotKey = "{{$HTMLPLACEHOLDER" + slotNum + "}}"
		}

		recursiveResult := e.mergeTemplateSlots(slotContent, appSite, appView, loader, enableJsonProcessing)
		recursiveResult = e.replaceTemplatePlaceholders(recursiveResult, appSite, appView, loader, enableJsonProcessing)
		slotContents[slotKey] = recursiveResult

		searchPos = closeStart + len(closeTag)
	}

	return slotContents
}

func (e *EngineNormalJson) replaceTemplatePlaceholders(html, appSite, appView string, loader interfaces.ILoaderJson, enableJsonProcessing bool) string {
	result := html
	searchPos := 0

	for searchPos < len(result) {
		openStart := strings.Index(result[searchPos:], "{{")
		if openStart == -1 {
			break
		}
		openStart += searchPos

		if openStart+2 < len(result) && (result[openStart+2] == '#' || result[openStart+2] == '@' || result[openStart+2] == '$' || result[openStart+2] == '/') {
			searchPos = openStart + 2
			continue
		}

		closeStart := strings.Index(result[openStart+2:], "}}")
		if closeStart == -1 {
			break
		}
		closeStart += openStart + 2

		placeholderName := strings.TrimSpace(result[openStart+2 : closeStart])
		if placeholderName == "" || !common.IsAlphaNumeric(placeholderName) {
			searchPos = openStart + 2
			continue
		}

		templateContent := e.getTemplateWithJson(appSite, placeholderName, appView, loader, enableJsonProcessing)

		if templateContent != "" {
			processedReplacement := e.replaceTemplatePlaceholders(templateContent, appSite, appView, loader, enableJsonProcessing)
			placeholder := result[openStart : closeStart+2]
			result = strings.ReplaceAll(result, placeholder, processedReplacement)
			searchPos = openStart + len(processedReplacement)
		} else {
			searchPos = closeStart + 2
		}
	}

	return result
}

File: csharp/Assembler/Engine/NormalJson/EngineNormalJson.cs
using Arshu.Common;
using Assembler.Common;
using Assembler.Interface;
using System.Collections.Generic;

namespace Assembler.Engine.NormalJson;

// Always use Base JsonObject/JsonArray types for consistent processing
/// <summary>
/// NEW Normal engine that uses ILoaderJson interface and JsonObject
/// Based on EngineNormal but with improved architecture and type safety
/// </summary>
public class EngineNormalJson
{
    #region Merge Templates

    public string AppViewPrefix { get; set; } = string.Empty;

    /// <summary>
    /// Merges templates by replacing placeholders with corresponding HTML
    /// This is a hybrid method that processes both slotted templates and simple placeholders
    /// JSON files with matching names are automatically merged with HTML templates before processing
    /// </summary>
    /// <param name="appSite">The application site name for template key generation</param>
    /// <param name="appView">The application view name (optional)</param>
    /// <param name="appFile">The application file name</param>
    /// <param name="loader">ILoaderJson instance providing templates and JSON merging</param>
    /// <param name="enableJsonProcessing">Whether to enable JSON data processing</param>
    /// <returns>HTML with placeholders replaced</returns>
    public string MergeTemplates(string appSite, string appFile, string? appView, ILoaderJson<string> loader, bool enableJsonProcessing = true)
    {
        Logger.Debug($"MergeTemplates called: appSite={appSite}, appFile={appFile}, appView={appView ?? "null"}, enableJson={enableJsonProcessing}", "EngineNormalJson");

        if (loader == null)
        {
            Logger.Warn("No loader provided", "EngineNormalJson");
            return "";
        }

        // Get main template using ILoaderJson (includes AppView fallback logic)
        var contentHtml = loader.GetTemplateHtml(appSite, appFile, appView, AppViewPrefix);
        if (string.IsNullOrEmpty(contentHtml))
        {
            Logger.Warn($"Main template not found for appSite={appSite}, appFile={appFile}", "EngineNormalJson");
            return string.Empty;
        }

        Logger.Debug($"Main template found, html size: {contentHtml.Length}", "EngineNormalJson");

        // Merge main template with JSON data using loader's centralized method
        if (enableJsonProcessing)
        {
            contentHtml = loader.MergeHtmlWithJson(contentHtml, appSite, appFile);
            Logger.Debug($"After main JSON merge: {contentHtml.Length} chars", "EngineNormalJson");
        }

        // Simple loop like Go implementation - avoid StringBuilder overhead
        // Templates are now loaded on-demand via GetTemplate method
        string previous;
        int maxPasses = 10;
        int actualPasses = 0;
        for (int pass = 0; pass < maxPasses; pass++)
        {
            previous = contentHtml;
            actualPasses = pass + 1;

            Logger.Debug($"Pass {actualPasses}, current size: {contentHtml.Length}", "EngineNormalJson");

            contentHtml = MergeTemplateSlots(contentHtml, appSite, appView, loader, enableJsonProcessing);
            Logger.Debug($"After slot merge: {contentHtml.Length} chars", "EngineNormalJson");

            contentHtml = ReplaceTemplatePlaceholders(contentHtml, appSite, appView, loader, enableJsonProcessing);
            Logger.Debug($"After placeholder replacement: {contentHtml.Length} chars", "EngineNormalJson");

            if (contentHtml == previous)
            {
                Logger.Debug($"No changes in pass {actualPasses}, stopping", "EngineNormalJson");
                break;
            }
        }

        Logger.Debug($"MergeTemplates complete after {actualPasses} passes: output size={contentHtml.Length}", "EngineNormalJson");
        return contentHtml;

    }

    #endregion

    #region Get Template (Private)

    /// <summary>
    /// Gets a template with on-demand loading and JSON merging from ILoaderJson
    /// </summary>
    private string? GetTemplateWithJson(string appSite, string templateName, ILoaderJson<string> loader, string? appView, bool enableJsonProcessing)
    {
        // Get HTML template
        var html = loader.GetTemplateHtml(appSite, templateName, appView, AppViewPrefix);
        if (string.IsNullOrEmpty(html))
            return null;

        Logger.Debug($"GetTemplateWithJson: template={templateName}, html size={html.Length}", "EngineNormalJson");

        // Merge with JSON if enabled using loader's centralized method
        if (enableJsonProcessing)
        {
            var originalSize = html.Length;
            html = loader.MergeHtmlWithJson(html, appSite, templateName);
            Logger.Debug($"After JSON merge for {templateName}: size {originalSize} -> {html.Length}", "EngineNormalJson");
        }

        return html;
    }

    #endregion

    #region Slot Processing (Private)

    /// <summary>
    /// IndexOf-based version: Recursively merges a slotted template (e.g., center.html, columns.html) with content.html
    /// Slot patterns in content.html: {{#TemplateName}} ... {{@HTMLPLACEHOLDER[N]}} ... {{/HTMLPLACEHOLDER[N]}} ... {{/TemplateName}}
    /// </summary>
    /// <param name="contentHtml">The content HTML containing slot patterns</param>
    /// <param name="appSite">The application site name for template key generation</param>
    /// <param name="appView">Optional AppView for fallback logic</param>
    /// <param name="loader">ILoaderJson instance for on-demand template loading</param>
    /// <param name="enableJsonProcessing">Whether to enable JSON data processing</param>
    /// <returns>Merged HTML with slots filled</returns>
    private string MergeTemplateSlots(string contentHtml, string appSite, string? appView, ILoaderJson<string> loader, bool enableJsonProcessing)
    {
        if (string.IsNullOrEmpty(contentHtml))
            return contentHtml;

        string previous;
        do
        {
            previous = contentHtml;
            contentHtml = ProcessTemplateSlots(contentHtml, appSite, appView, loader, enableJsonProcessing);
        } while (contentHtml != previous);
        return contentHtml;
    }

    /// <summary>
    /// Helper method to process slotted templates using IndexOf
    /// </summary>
    private string ProcessTemplateSlots(string contentHtml, string appSite, string? appView, ILoaderJson<string> loader, bool enableJsonProcessing)
    {
        var result = contentHtml;
        var searchPos = 0;

        while (searchPos < result.Length)
        {
            // Look for opening tag {{#
            var openStart = result.IndexOf("{{#", searchPos);
            if (openStart == -1) break;

            // Find the end of the template name
            var openEnd = result.IndexOf("}}", openStart + 3);
            if (openEnd == -1) break;

            // Extract template name
            var templateName = result.Substring(openStart + 3, openEnd - openStart - 3).Trim();
            if (string.IsNullOrEmpty(templateName) || !CommonUtil.IsAlphaNumeric(templateName))
            {
                searchPos = openStart + 1;
                continue;
            }

            // Look for corresponding closing tag
            var closeTag = "{{/" + templateName + "}}";
            var closeStart = CommonUtil.FindMatchingCloseTag(result, openEnd + 2, "{{#" + templateName + "}}", closeTag);
            if (closeStart == -1)
            {
                searchPos = openStart + 1;
                continue;
            }

            // Extract inner content
            var innerStart = openEnd + 2;
            var innerContent = result.Substring(innerStart, closeStart - innerStart);

            // Load template with JSON on-demand
            var templateHtml = GetTemplateWithJson(appSite, templateName, loader, appView, enableJsonProcessing);

            if (!string.IsNullOrEmpty(templateHtml))
            {
                // Extract slot contents
                var slotContents = ExtractSlotContents(innerContent, appSite, appView, loader, enableJsonProcessing);

                // Replace slots in template
                var processedTemplate = templateHtml;
                foreach (var kvp in slotContents)
                {
                    processedTemplate = processedTemplate.Replace(kvp.Key, kvp.Value);
                }

                // Remove any remaining slot placeholders
                processedTemplate = CommonUtil.RemoveRemainingSlotPlaceholders(processedTemplate);

                // Replace the entire slotted section
                var fullMatch = result.Substring(openStart, closeStart + closeTag.Length - openStart);
                result = result.Replace(fullMatch, processedTemplate);
                searchPos = openStart + processedTemplate.Length;
            }
            else
            {
                searchPos = openStart + 1;
            }
        }

        return result;
    }

    /// <summary>
    /// Extract slot contents using IndexOf approach
    /// </summary>
    private Dictionary<string, string> ExtractSlotContents(string innerContent, string appSite, string? appView, ILoaderJson<string> loader, bool enableJsonProcessing)
    {
        var slotContents = new Dictionary<string, string>();
        var searchPos = 0;

        while (searchPos < innerContent.Length)
        {
            // Look for slot start {{@HTMLPLACEHOLDER
            var slotStart = innerContent.IndexOf("{{@HTMLPLACEHOLDER", searchPos);
            if (slotStart == -1) break;

            // Find the number (if any) and closing }}
            var afterPlaceholder = slotStart + 18; // Length of "{{@HTMLPLACEHOLDER"
            var slotNum = "";
            var pos = afterPlaceholder;

            // Extract slot number
            while (pos < innerContent.Length && char.IsDigit(innerContent[pos]))
            {
                slotNum += innerContent[pos];
                pos++;
            }

            // Check for closing }}
            if (pos + 1 >= innerContent.Length || innerContent.Substring(pos, 2) != "}}")
            {
                searchPos = slotStart + 1;
                continue;
            }

            var slotOpenEnd = pos + 2;

            // Find matching closing tag
            var closeTag = string.IsNullOrEmpty(slotNum) ? "{{/HTMLPLACEHOLDER}}" : $"{{{{/HTMLPLACEHOLDER{slotNum}}}}}";
            var openTag = string.IsNullOrEmpty(slotNum) ? "{{@HTMLPLACEHOLDER}}" : $"{{{{@HTMLPLACEHOLDER{slotNum}}}}}";

            var closeStart = CommonUtil.FindMatchingCloseTag(innerContent, slotOpenEnd, openTag, closeTag);
            if (closeStart == -1)
            {
                searchPos = slotStart + 1;
                continue;
            }

            // Extract slot content
            var slotContent = innerContent.Substring(slotOpenEnd, closeStart - slotOpenEnd);

            // Generate slot key
            var slotKey = string.IsNullOrEmpty(slotNum) ? "{{$HTMLPLACEHOLDER}}" : $"{{{{$HTMLPLACEHOLDER{slotNum}}}}}";

            // Process both slotted templates AND simple placeholders in slot content
            // This enables proper nested template processing to match the preprocessing implementation
            var recursiveResult = MergeTemplateSlots(slotContent, appSite, appView, loader, enableJsonProcessing);
            recursiveResult = ReplaceTemplatePlaceholders(recursiveResult, appSite, appView, loader, enableJsonProcessing);
            slotContents[slotKey] = recursiveResult;

            searchPos = closeStart + closeTag.Length;
        }

        return slotContents;
    }

    #endregion

    #region PlaceHolder Processing (Private)

    /// <summary>
    /// Helper method to process simple placeholders only (without slotted template processing)
    /// </summary>
    private string ReplaceTemplatePlaceholders(string html, string appSite, string? appView, ILoaderJson<string> loader, bool enableJsonProcessing)
    {
        var result = html;
        var searchPos = 0;

        while (searchPos < result.Length)
        {
            // Look for opening placeholder {{
            var openStart = result.IndexOf("{{", searchPos);
            if (openStart == -1) break;

            // Make sure it's not a slotted template or special placeholder
            if (openStart + 2 < result.Length && (result[openStart + 2] == '#' || result[openStart + 2] == '@' || result[openStart + 2] == '$' || result[openStart + 2] == '/'))
            {
                searchPos = openStart + 2;
                continue;
            }

            // Find closing }}
            var closeStart = result.IndexOf("}}", openStart + 2);
            if (closeStart == -1) break;

            // Extract placeholder name
            var placeholderName = result.Substring(openStart + 2, closeStart - openStart - 2).Trim();
            if (string.IsNullOrEmpty(placeholderName) || !CommonUtil.IsAlphaNumeric(placeholderName))
            {
                searchPos = openStart + 2;
                continue;
            }

            // Load template with JSON on-demand
            var templateContent = GetTemplateWithJson(appSite, placeholderName, loader, appView, enableJsonProcessing);

            if (!string.IsNullOrEmpty(templateContent))
            {
                // Recursively process the loaded template
                var processedReplacement = ReplaceTemplatePlaceholders(templateContent, appSite, appView, loader, enableJsonProcessing);
                var placeholder = result.Substring(openStart, closeStart + 2 - openStart);
                result = result.Replace(placeholder, processedReplacement);
                searchPos = openStart + processedReplacement.Length;
            }
            else
            {
                searchPos = closeStart + 2;
            }
        }

        return result;
    }

    #endregion
}
